Support for scale

andrcuns opened the pull request:
This implements issue #37 
Both `--scale` option as well as added parameter `scale:` in compose file version 2.2 are supported.

The scale does introduce a new concept where multiple instances of one service might exist and therefore naming exactly how compose names containers have to be used. That directly impacts both exposed environment variable and system property names and getting serviceInfos by name.
I would like to hear Your opinion on how to deal with this since at the current state this PR isn't backwards compatible with service naming used in mentioned functionality.

augi commented on src/main/groovy/com/avast/gradle/dockercompose/tasks/ComposeUp.groovy:
```diff
@@ -98,17 +104,20 @@ class ComposeUp extends DefaultTask {
+            ServiceHost host = getServiceHost(serviceName, inspection)
+            logger.info("Will use $host as host of service $serviceName")
+            def tcpPorts = getTcpPortsMapping(serviceName, inspection, host)
+            new ServiceInfo(name: inspection.Name.replace('/', ''), serviceHost: host, tcpPorts: tcpPorts, containerHostname: inspection.Config.Hostname, inspection: inspection)
```
I would prefer to replace also the project name, like transform `test_web_1` to `web_1`. The project name is name of the directory but it can also be set by `ComposeExtension.projectName`.

augi replied:
Actually, this is little bit misleading. I would prefer to add a new class `ServiceInstanceInfo`. `ServiceInfo` would contain collection of `ServiceInstanceInfo`.

augi replied:
I know it's not so easy and so I could prepare my PR based on the yours. Do you agree?

andrcuns replied:
Yes, I'm fine with that.

andrcuns replied:
Had a free evening and tried to implement requested changes, but still feel free to prepare your own PR if necessary.

augi commented on src/main/groovy/com/avast/gradle/dockercompose/ServiceInfo.groovy:
```diff
@@ -1,20 +1,10 @@
 
-import groovy.transform.Immutable
-
-@Immutable
```
I would prefer if the class remained immutable, please.

augi commented on src/main/groovy/com/avast/gradle/dockercompose/ServiceInfo.groovy:
```diff
@@ -1,20 +1,10 @@
-    Map<String, Object> inspection
-    String getContainerId() { inspection.Id }
-
-    String getHost() { serviceHost.host }
```
For sake of backward compatibility, what about keeping these methods so they would proxy to the first instance? Like `String getHost() { serviceInstanceInfos[0].getHost() }`

But maybe nobody uses these methods so keep it as is.

augi commented on src/main/groovy/com/avast/gradle/dockercompose/tasks/ComposeUp.groovy:
```diff
@@ -102,13 +109,19 @@ class ComposeUp extends DefaultTask {
-        new ServiceInfo(name: serviceName, serviceHost: host, tcpPorts: tcpPorts, containerHostname: inspection.Config.Hostname, inspection: inspection)
+        Iterable<String> containerIds = getContainerIds(serviceName)
+        ServiceInfo serviceInfo = new ServiceInfo(name: serviceName)
+        containerIds.each { String containerId ->
```
Please extract this lambda to `createServiceInstanceInfo` method. I believe it will be more readable.

augi commented on src/main/groovy/com/avast/gradle/dockercompose/tasks/ComposeUp.groovy:
```diff
@@ -102,13 +109,19 @@ class ComposeUp extends DefaultTask {
+            ServiceHost host = getServiceHost(serviceName, inspection)
+            logger.info("Will use $host as host of service $serviceName")
+            def tcpPorts = getTcpPortsMapping(serviceName, inspection, host)
+            ServiceInstanceInfo serviceInstanceInfo = new ServiceInstanceInfo(instanceName: inspection.Name.find(/${serviceName}_\d+/), 
```
Perfect! I like how you are extracting the `instanceName` :)

augi commented on src/main/groovy/com/avast/gradle/dockercompose/tasks/ComposeUp.groovy:
```diff
@@ -102,13 +109,19 @@ class ComposeUp extends DefaultTask {
-        def tcpPorts = getTcpPortsMapping(serviceName, inspection, host)
-        new ServiceInfo(name: serviceName, serviceHost: host, tcpPorts: tcpPorts, containerHostname: inspection.Config.Hostname, inspection: inspection)
+        Iterable<String> containerIds = getContainerIds(serviceName)
+        ServiceInfo serviceInfo = new ServiceInfo(name: serviceName)
```
As the `ServiceInfo` will be immutable, you will have to move this constructor call to the end of this method.

augi commented on src/main/groovy/com/avast/gradle/dockercompose/ComposeExtension.groovy:
```diff
@@ -62,21 +63,25 @@ class ComposeExtension {
-            task.environment.put("${si.name.toUpperCase()}_CONTAINER_HOSTNAME".toString(), si.containerHostname)
-            si.tcpPorts.each {
-                task.environment.put("${si.name.toUpperCase()}_TCP_${it.key}".toString(), it.value)
+        servicesInfos.values().each { serviceInfo ->
```
It would be great to make this change in both `exposeAs` methods: If the `instanceName` ends with `_1` then add envvar/sysprop that will have name without this suffix. So we will expose both `web_1` and `web`.

augi replied:
Also please document this behavior in README.

augi commented on src/main/groovy/com/avast/gradle/dockercompose/ComposeExtension.groovy:
```diff
@@ -125,6 +130,10 @@ class ComposeExtension {
     }
+    
+    boolean scale() {
+        getDockerComposeVersion() >= VersionNumber.parse('1.13.0') && this.scale
```
We should throw an exception if somebody set `scale` but they don't have Docker Compose >= 1.13.0.

andrcuns replied:
I will address the suggestions tomorrow during the day. Thanks for the feedback.

andrcuns commented:
I tried to address all of the requested changes and added additional testing with compose version 1.13.0, but wasn't sure which variations of versions You would like to test with and if it is ok for tests rely on passing compose version as environment variable since it was the only convenient way I could think of on how to ignore certain tests on different compose versions.
Let me know what do You think.

augi commented on README.md:
```diff
@@ -57,17 +57,22 @@ dockerCompose {
     // dockerComposeWorkingDirectory = '/path/where/docker-compose/is/invoked/from'
     // dockerComposeStopTimeout = java.time.Duration.ofSeconds(20) // time before docker-compose sends SIGTERM to the running containers after the composeDown task has been started
     // environment.put 'BACKEND_ADDRESS', '192.168.1.100' // Pass environment variable to 'docker-compose' for substitution in compose file
+    // scale = [${serviceName1}: 5, ${serviceName2}: 2] // Pass docker compose --scale option like 'docker-compose up --scale SERVICE1=NUM --scale SERVICE2=NUM'
```
Please change to this:
`// Pass docker compose --scale option like 'docker-compose up --scale serviceName1=5 --scale serviceName2=2'`

augi commented on src/main/groovy/com/avast/gradle/dockercompose/ServiceInfo.groovy:
```diff
@@ -5,16 +5,14 @@ import groovy.transform.Immutable
-    /* Docker inspection */
-    Map<String, Object> inspection
-    String getContainerId() { inspection.Id }
+    List<ServiceInstanceInfo> serviceInstanceInfos = []
```
What about changing it from `List<ServiceInstanceInfo>` to `Map<String, ServiceInstanceInfo>`? It would simplify the usage on few places (and also in the public API as described in README).

andrcuns replied:
And the key would be instance name, right?

augi replied:
Right.

andrcuns commented:
Addressed the last 2 comments and also added method getFirstInstance to save people from typing serviceInstaneInfos.values().first() in case of one container running.

augi commented on src/main/groovy/com/avast/gradle/dockercompose/ServiceInfo.groovy:
```diff
@@ -5,16 +5,18 @@ import groovy.transform.Immutable
 
-    String getHost() { serviceHost.host }
-    Map<Integer, Integer> getPorts() { tcpPorts }
+    String getHost() { serviceInstanceInfos.values().first().serviceHost.host }
```
Please use getFirstInstance method here.

augi commented:
Perfect, I'll release it in few hours!
